Jellyfin.Plugin.OpenSubtitles/OpenSubtitleDownloader.cs

```diff
@@ -207,124 +168,143 @@ await Task.Delay(250, cancellationToken)
             var ossId = idParts[2];
 
-            var downloadsList = new[] { int.Parse(ossId, UsCulture) };
+            var fid = int.Parse(ossId, _usCulture);
 
-            await Login(cancellationToken).ConfigureAwait(false);
+            var info = await OpenSubtitlesHandler.OpenSubtitles.GetSubtitleLinkAsync(fid, _login, cancellationToken).ConfigureAwait(false);
 
-            if ((DateTime.UtcNow - _lastRateLimitException).TotalHours < 1)
+            if (!info.Ok)
             {
-                throw new RateLimitExceededException("OpenSubtitles rate limit reached");
-            }
+                switch (info.Code)
+                {
+                    case HttpStatusCode.NotAcceptable when info.Data.Remaining <= 0:
+                    {
+                        if (_login?.User != null)
+                        {
+                            _login.User.RemainingDownloads = 0;
+                        }
 
-            if (_rateLimitLeft == 0)
-            {
-                await Task.Delay(1000, cancellationToken)
-                    .ConfigureAwait(false);
-            }
+                        throw new OpenApiException("OpenSubtitles download limit reached");
+                    }
 
-            var downloadResponse = await OpenSubtitlesHandler.OpenSubtitles.DownloadSubtitlesAsync(downloadsList, cancellationToken).ConfigureAwait(false);
+                    case HttpStatusCode.Unauthorized:
+                        // JWT token expired, obtain a new one and try again?
+                        _login = null;
+                        return await GetSubtitlesInternal(id, cancellationToken).ConfigureAwait(false);
+                }
 
-            var downloadResult = downloadResponse.Item1;
-            _rateLimitLeft = downloadResponse.Item2 ?? _rateLimitLeft;
+                var msg = info.Body.Contains("<html", StringComparison.OrdinalIgnoreCase) ? "[html]" : info.Body;
 
-            if (downloadResponse.Item2 != null)
-            {
-                if (_rateLimitLeft <= 4)
-                {
-                    await Task.Delay(250, cancellationToken)
-                        .ConfigureAwait(false);
-                }
+                msg = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "Invalid response for file {0}: {1}\n\n{2}",
+                    fid,
+                    info.Code,
+                    msg);
+
+                throw new OpenApiException(msg);
             }
 
-            if ((downloadResult.Status ?? string.Empty).IndexOf("407", StringComparison.OrdinalIgnoreCase) != -1)
+            if (_login?.User != null)
             {
-                _lastRateLimitException = DateTime.UtcNow;
-                throw new RateLimitExceededException("OpenSubtitles daily limit reached");
+                _login.User.RemainingDownloads = info.Data.Remaining;
+                _logger.LogInformation($"Remaining downloads: {_login.User.RemainingDownloads}");
```

Comment:
Please let the logger handle the string formatting. This change request is not specific to this line, but applies to all of them.
```suggestion
                _logger.LogInformation("Remaining downloads: {RemainingDownloads}", _login.User.RemainingDownloads);
```

Reply:
done in b6f9b530fd999ae6b9b21ababef5b019f4c921fb